### PR TITLE
DATACASS-155 - Support TIMESTAMP in WriteOptions

### PR DESCRIPTION
**Pull request by lukasz-antoniak**
JIRA ticket: https://jira.spring.io/browse/DATACASS-155

### CODE REVIEW

**Comment by pivotal-issuemaster** [comment]
@lukasz-antoniak Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-projects/spring-data-cassandra&pullRequestId=122)!

[Click here](https://cla.pivotal.io/sync/spring?repositoryId=spring-projects/spring-data-cassandra&pullRequestId=122) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/about) for frequently asked questions.

**Comment by pivotal-issuemaster** [comment]
@lukasz-antoniak Thank you for signing the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-projects/spring-data-cassandra&pullRequestId=122)!

File: spring-data-cassandra/src/main/java/org/springframework/data/cassandra/core/cql/WriteOptions.java [hunk 1]
```diff
@@ -40,6 +41,7 @@
 	private static final WriteOptions EMPTY = new WriteOptionsBuilder().build();
 
 	private final Duration ttl;
+	private final long timestamp;
```
Comment (mp911de) [inline spring-data-cassandra/src/main/java/org/springframework/data/cassandra/core/cql/WriteOptions.java]: This one should rather be `@Nullable Long timestamp`

File: spring-data-cassandra/src/main/java/org/springframework/data/cassandra/core/cql/QueryOptionsUtil.java [hunk 2]
```diff
@@ -106,6 +106,9 @@ public static Insert addWriteOptions(Insert insert, WriteOptions writeOptions) {
 		if (!writeOptions.getTtl().isNegative()) {
 			insert.using(QueryBuilder.ttl(Math.toIntExact(writeOptions.getTtl().getSeconds())));
 		}
+		if (!(writeOptions.getTimestamp() < 0)) {
```
Comment (mp911de) [inline spring-data-cassandra/src/main/java/org/springframework/data/cassandra/core/cql/QueryOptionsUtil.java]: `WriteOptions` (and `QueryOptions` in general) uses nullable properties to indicate presence/absence of options.

File: spring-data-cassandra/src/main/java/org/springframework/data/cassandra/core/cql/QueryOptionsUtil.java [hunk 3]
```diff
@@ -126,6 +129,9 @@ public static Update addWriteOptions(Update update, WriteOptions writeOptions) {
 		if (!writeOptions.getTtl().isNegative()) {
 			update.using(QueryBuilder.ttl(Math.toIntExact(writeOptions.getTtl().getSeconds())));
 		}
+		if (!(writeOptions.getTimestamp() < 0)) {
```
Comment (mp911de) [inline spring-data-cassandra/src/main/java/org/springframework/data/cassandra/core/cql/QueryOptionsUtil.java]: `WriteOptions` (and `QueryOptions` in general) uses nullable properties to indicate presence/absence of options.

File: spring-data-cassandra/src/main/java/org/springframework/data/cassandra/core/cql/WriteOptions.java [hunk 4]
```diff
@@ -68,14 +70,16 @@ public WriteOptions(@Nullable ConsistencyLevel consistencyLevel, @Nullable Retry
 		super(consistencyLevel, retryPolicy);
 
 		this.ttl = ttl == null ? Duration.ofMillis(-1) : Duration.ofSeconds(ttl);
+		this.timestamp = -1;
 	}
 
 	protected WriteOptions(@Nullable ConsistencyLevel consistencyLevel, @Nullable RetryPolicy retryPolicy,
-			@Nullable Boolean tracing, @Nullable Integer fetchSize, Duration readTimeout, Duration ttl) {
+			@Nullable Boolean tracing, @Nullable Integer fetchSize, Duration readTimeout, Duration ttl, long timestamp) {
```
Comment (mp911de) [inline spring-data-cassandra/src/main/java/org/springframework/data/cassandra/core/cql/WriteOptions.java]: `timestamp` should use boxed `Long`.

File: spring-data-cassandra/src/main/java/org/springframework/data/cassandra/core/UpdateOptions.java [hunk 5]
```diff
@@ -40,9 +41,9 @@
 	private boolean ifExists;
 
 	private UpdateOptions(@Nullable ConsistencyLevel consistencyLevel, @Nullable RetryPolicy retryPolicy,
-			@Nullable Boolean tracing, @Nullable Integer fetchSize, Duration readTimeout, Duration ttl, boolean ifExists) {
+			@Nullable Boolean tracing, @Nullable Integer fetchSize, Duration readTimeout, Duration ttl, long timestamp, boolean ifExists) {
```
Comment (mp911de) [inline spring-data-cassandra/src/main/java/org/springframework/data/cassandra/core/UpdateOptions.java]: `timestamp` should use boxed `Long`.

File: spring-data-cassandra/src/main/java/org/springframework/data/cassandra/core/cql/WriteOptions.java [hunk 6]
```diff
@@ -125,6 +136,7 @@ public Duration getTtl() {
 	public static class WriteOptionsBuilder extends QueryOptionsBuilder {
 
 		protected Duration ttl = Duration.ofMillis(-1);
+		protected long timestamp = -1;
```
Comment (mp911de) [inline spring-data-cassandra/src/main/java/org/springframework/data/cassandra/core/cql/WriteOptions.java]: `timestamp` should use boxed `Long`.

File: spring-data-cassandra/src/main/java/org/springframework/data/cassandra/core/cql/WriteOptions.java [hunk 7]
```diff
@@ -236,14 +249,38 @@ public WriteOptionsBuilder ttl(Duration ttl) {
...
+
+		/**
+		 * Sets the timestamp of write operations.
+		 *
+		 * @param timestamp mutation date time.
+		 * @return {@code this} {@link WriteOptionsBuilder}
+		 */
+		public WriteOptionsBuilder timestamp(Instant timestamp) {
+			Assert.notNull(timestamp, "Timestamp must not be null");
+			this.timestamp = timestamp.toEpochMilli() * 1000;
```
Comment (mp911de) [inline spring-data-cassandra/src/main/java/org/springframework/data/cassandra/core/cql/WriteOptions.java]: How about `TimeUnit.MILLISECONDS.toMicros(…)`?

File: spring-data-cassandra/src/test/java/org/springframework/data/cassandra/core/InsertOptionsUnitTests.java [hunk 8]
```diff
@@ -45,14 +51,16 @@ public void buildInsertOptionsMutate() {
...
 				.withIfNotExists()
 				.build();
 
-		InsertOptions mutated = insertOptions.mutate().ttl(Duration.ofSeconds(5)).build();
+		InsertOptions mutated = insertOptions.mutate().ttl(Duration.ofSeconds(5)).timestamp(1519200753).build();
 
 		assertThat(mutated).isNotNull();
 		assertThat(mutated).isNotSameAs(insertOptions);
 		assertThat(mutated.getTtl()).isEqualTo(Duration.ofSeconds(5));
+		assertThat(mutated.getTimestamp() == 1519200753);
```
Comment (mp911de) [inline spring-data-cassandra/src/test/java/org/springframework/data/cassandra/core/InsertOptionsUnitTests.java]: did you mean `assertThat(mutated.getTimestamp()).isEqualTo(1519200753)`?

File: spring-data-cassandra/src/test/java/org/springframework/data/cassandra/core/UpdateOptionsUnitTests.java [hunk 9]
```diff
@@ -30,13 +33,16 @@
...
+		Instant now = LocalDateTime.now().toInstant(ZoneOffset.UTC);
 
 		UpdateOptions updateOptions = UpdateOptions.builder()
 				.ttl(10)
+				.timestamp(now)
 				.withIfExists()
 				.build();
 
 		assertThat(updateOptions.getTtl()).isEqualTo(Duration.ofSeconds(10));
+		assertThat(updateOptions.getTimestamp() == now.toEpochMilli() * 1000);
```
Comment (mp911de) [inline spring-data-cassandra/src/test/java/org/springframework/data/cassandra/core/UpdateOptionsUnitTests.java]: did you mean assertThat(mutated.getTimestamp()).isEqualTo(now.toEpochMilli() * 1000)?

File: spring-data-cassandra/src/test/java/org/springframework/data/cassandra/core/UpdateOptionsUnitTests.java [hunk 10]
```diff
@@ -45,14 +51,16 @@ public void buildUpdateOptionsMutate() {
...
 				.withIfExists()
 				.build();
 
-		UpdateOptions mutated = updateOptions.mutate().ttl(20).build();
+		UpdateOptions mutated = updateOptions.mutate().ttl(20).timestamp(1519000753).build();
 
 		assertThat(mutated).isNotNull();
 		assertThat(mutated).isNotSameAs(updateOptions);
 		assertThat(mutated.getTtl()).isEqualTo(Duration.ofSeconds(20));
+		assertThat(mutated.getTimestamp() == 1519000753);
```
Comment (mp911de) [inline spring-data-cassandra/src/test/java/org/springframework/data/cassandra/core/UpdateOptionsUnitTests.java]: See above.

File: spring-data-cassandra/src/test/java/org/springframework/data/cassandra/core/InsertOptionsUnitTests.java [hunk 11]
```diff
@@ -30,13 +33,16 @@
...
+		Instant now = LocalDateTime.now().toInstant(ZoneOffset.UTC);
 
 		InsertOptions insertOptions = InsertOptions.builder()
 				.ttl(10)
+				.timestamp(now)
 				.withIfNotExists()
 				.build();
 
 		assertThat(insertOptions.getTtl()).isEqualTo(Duration.ofSeconds(10));
+		assertThat(insertOptions.getTimestamp() == now.toEpochMilli() * 1000);
```
Comment (mp911de) [inline spring-data-cassandra/src/test/java/org/springframework/data/cassandra/core/InsertOptionsUnitTests.java]: did you mean `assertThat(insertOptions.getTimestamp()).isEqualTo(now.toEpochMilli() * 1000);`?

File: spring-data-cassandra/src/test/java/org/springframework/data/cassandra/core/cql/WriteOptionsUnitTests.java [hunk 12]
```diff
@@ -38,13 +41,15 @@ public void buildWriteOptions() {
...
 				.ttl(123)
+				.timestamp(1519000753)
 				.retryPolicy(FallthroughRetryPolicy.INSTANCE)
 				.readTimeout(1)
 				.fetchSize(10)
 				.withTracing()
 				.build();
 
 		assertThat(writeOptions.getTtl()).isEqualTo(Duration.ofSeconds(123));
+		assertThat(writeOptions.getTimestamp() == 1519000753);
```
Comment (mp911de) [inline spring-data-cassandra/src/test/java/org/springframework/data/cassandra/core/cql/WriteOptionsUnitTests.java]: see above

File: spring-data-cassandra/src/test/java/org/springframework/data/cassandra/core/cql/WriteOptionsUnitTests.java [hunk 13]
```diff
@@ -95,6 +102,7 @@ public void buildWriteOptionsMutate() {
 		assertThat(mutated).isNotNull();
 		assertThat(mutated).isNotSameAs(writeOptions);
 		assertThat(mutated.getTtl()).isEqualTo(Duration.ofSeconds(123));
+		assertThat(mutated.getTimestamp() == now.toEpochMilli() * 1000);
```
Comment (mp911de) [inline spring-data-cassandra/src/test/java/org/springframework/data/cassandra/core/cql/WriteOptionsUnitTests.java]: see above

**Comment by lukasz-antoniak** [comment]
Thank you for quick review. I have applied the changes and squashed commits. Hope everything is fine now. This was just a warm-up and my first contribution to Spring Data project :). Apologies for mistake in unit tests. I did not use AssertJ before and JUnit habits came in (same `assertThat` method name).

**Comment by mp911de** [comment]
Awesome! No worries, we're happy to guide you through pull requests. I think we'll be able to do the merge within the next days.

**Comment by mp911de** [comment]
That's merged and polished now.

**Pull request closed without merging**